src/common-libs/plan/index.ts

```diff
@@ -0,0 +1,511 @@
+  planTag: AddPlanTagInput
+) => {
+  const reportingPeriods = await models.planReportingPeriod.find({
+    where: {
+      planId: createBrandedValue(planTag.planId),
+    },
+    orderBy: { column: 'periodNumber', order: 'asc' },
+  });
+  if (!reportingPeriods.length) {
+    throw new Error(
+      'Cannot activate planTag if associated plan has no reporting period'
+    );
+  }
+  const latestPlanVersion = await models.planVersion.findOne({
+    where: {
+      planId: createBrandedValue(planTag.planId),
+      latestVersion: true,
+    },
+  });
+  if (latestPlanVersion && !latestPlanVersion.currentReportingPeriodId) {
+    await models.planVersion.update({
+      values: {
+        currentReportingPeriodId: reportingPeriods[0].id,
+      },
+      where: {
+        id: latestPlanVersion.id,
+      },
+    });
+  }
+};
+
+const getAssociatedPlanVersion = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>
+) => {
+  const associatedPlanVersion = await models.planVersion.findOne({
+    where: {
+      planId: planTag.planId,
+      createdAt: {
+        [models.Op.LTE]: planTag.createdAt,
+      },
+      latestVersion: true,
+    },
+  });
+  if (!associatedPlanVersion) {
+    throw new Error('Cannot activate planTag if no associated version exist');
+  }
+  return associatedPlanVersion;
+};
+
+export const updateLastPublishedReportingPeriodId = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>
+) => {
+  const reportingPeriods = await models.planReportingPeriod.find({
+    where: {
+      planId: planTag.planId,
+      measurementsGenerated: true,
+    },
+    orderBy: { column: 'periodNumber', order: 'desc' },
+  });
+
+  if (reportingPeriods.length) {
+    const latestReportingPeriodId = reportingPeriods[0].id;
+    const publishedMeasurements = await models.measurement.find({
+      where: {
+        planReportingPeriodId: latestReportingPeriodId,
+        versionTags: { [models.Op.NOT_IN]: [['{}']] },
+      },
+    });
+    console.log(publishedMeasurements.length);
+    if (publishedMeasurements.length) {
+      const associatedPlanVersion = await getAssociatedPlanVersion(
+        models,
+        planTag
+      );
+      console.log(associatedPlanVersion);
+      await models.planVersion.update({
+        values: {
+          lastPublishedReportingPeriodId: latestReportingPeriodId,
+        },
+        where: {
+          id: associatedPlanVersion.id,
+        },
+      });
+    }
+  }
+};
+
+const versionedModels: any = [
+  { table: 'attachment', relationship: true },
+  { table: 'governingEntity', relationship: true },
+  { table: 'planEntity', relationship: true },
+  { table: 'planLocation' },
+  { table: 'planVersion' },
+  { table: 'planYear' },
+];
+export const updateVersionStates = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>,
+  publishMeasurements: boolean,
+  planReportingPeriodIds: PlanReportingPeriodId[] | []
```

Comment:
What does this mean?
```suggestion
  planReportingPeriodIds: PlanReportingPeriodId[]
```